### PR TITLE
tools: check for unstaged changes before formatting

### LINKED ISSUE #302: make format should check for unstaged changes
Running `make format` overwrites files in place, which I think is the right behavior, but it could surprise some folks. It's nice to see what it changed and to a quick sanity check over the changes.

I imagine the best workflow would be:

  1. make changes
  2. git add all changes
  3. run format
  4. check over format
  5. git add format results
  6. git commit

which we can help guide people to with some niceties

### PR DESCRIPTION
**Pull request by ppannuto**
Closes #302.

### CODE REVIEW

File: tools/run_cargo_fmt.sh [hunk 1]
```diff
@@ -6,6 +6,28 @@
...
+# `git status --porcelain` formats things for scripting
+# | M changed file, unstaged
+# |M  changed file, staged (git add has run)
+# |MM changed file, some staged and some unstaged changes (git add then changes)
+# |?? untracked file
+if git status --porcelain | grep '^.M'; then
+	echo "$(tput bold)Warning: Formatting will overwrite files in place.$(tput sgr0)"
+	echo "While this is probably what you want, it's often useful to"
+	echo "stage all of your changes (git add ...) before format runs,"
+	echo "just so you can double-check everything"
```
Comment (bradjc) [inline tools/run_cargo_fmt.sh]: Add `.`

File: tools/run_cargo_fmt.sh [hunk 2]
```diff
@@ -6,6 +6,28 @@
...
 
+# Format overwrites changes, which is probably good, but it's nice to see
+# what it has done
+#
+# `git status --porcelain` formats things for scripting
+# | M changed file, unstaged
+# |M  changed file, staged (git add has run)
+# |MM changed file, some staged and some unstaged changes (git add then changes)
+# |?? untracked file
+if git status --porcelain | grep '^.M'; then
```
Comment (bradjc) [inline tools/run_cargo_fmt.sh]: What about

    if git status --porcelain | grep '^.M.*\.rs' > /dev/null; then

That checks that the changed files are .rs files only, and hides the git status before the actual message.

File: tools/run_cargo_fmt.sh [hunk 3]
```diff
@@ -6,6 +6,28 @@
...
+	echo "$(tput bold)Warning: Formatting will overwrite files in place.$(tput sgr0)"
+	echo "While this is probably what you want, it's often useful to"
+	echo "stage all of your changes (git add ...) before format runs,"
+	echo "just so you can double-check everything"
+	echo ""
+	echo "$(tput bold)git status:$(tput sgr0)"
+	git status
+	echo ""
+	read -p "Continue formatting with a dirty tree? [y/N] " response
```
Comment (bradjc) [inline tools/run_cargo_fmt.sh]: I think this would be better as "Continue formatting with unstaged changes?"

**Comment by ppannuto** [comment]
fixed

File: userland/tools/uncrustify/uncrustify.sh [hunk 4]
```diff
@@ -2,6 +2,29 @@
...
 
+# Format overwrites changes, which is probably good, but it's nice to see
+# what it has done
+#
+# `git status --porcelain` formats things for scripting
+# | M changed file, unstaged
+# |M  changed file, staged (git add has run)
+# |MM changed file, some staged and some unstaged changes (git add then changes)
+# |?? untracked file
+if git status --porcelain | grep '^.M.*\.rs' -q; then
```
Comment (bradjc) [inline userland/tools/uncrustify/uncrustify.sh]: Shouldn't this look for `.c` and `.h` files?

**Comment by ppannuto** [comment]
Yes, good catch. New regex should catch an of the .c .cpp .cxx .cc .h .hpp family of things

**Pull request merged**